akka-http-core/src/main/scala/akka/http/impl/engine/server/ServerTerminator.scala

```diff
@@ -216,6 +221,11 @@ private[http] final class GracefulTerminatorStage(settings: ServerSettings)
           pendingUserHandlerResponse = true
           push(toUser, request)
         }
+
+        override def onUpstreamFinish(): Unit = {
+          // don't finish the whole bidi stage, just propagate the completion:
+          complete(toUser)
+        }
```

Comment:
Do we have to keep track and stop the stage when both flows have completed?

Reply:
That was my first thought as well, if we keep the bidistage alive here, then when _does_ it actually stop

Reply:
Maybe do some debug printing in the stop callback and verify that something else makes it stop?

Reply:
`postStop` seems to be called fine. Also https://doc.akka.io/docs/akka/current/stream/stream-customize.html#custom-processing-with-graphstage says "`completeStage()` is equivalent to closing all output ports and cancelling all input ports.", it sounds like this should be sufficient.

Reply:
Thanks for verifying!